### PR TITLE
[SPARK-36620][SHUFFLE] Add client side push based shuffle metrics

### PR DESCRIPTION
**Pull request by thejdeep**
### What changes were proposed in this pull request?
This adds the following push based shuffle metrics like :
- Merger count and magnet enabled/disabled for stage
- Time spent on results finalization
- Counting actual number of blocks other than chunks
- Corrupt shuffle blocks chunks and fallback

 ### Why are the changes needed?
These changes help to understand the push based shuffle metrics of the application

 ### Does this PR introduce _any_ user-facing change?
Changes to API responses by SHS (eg: /stages)

 ### How was this patch tested?
Modified existing unit tests and also tested API response on event log files in SHS

Lead-authored by: Venkata krishnan Sowrirajan <vsowrirajan@linkedin.com>
Co-authored by: Chandni Singh <chsingh@linkedin.com>
Co-authored by: Thejdeep Gudivada <tgudivada@linkedin.com>

### CODE REVIEW

**Comment by mridulm** [comment]
Ok to test.

File: core/src/test/scala/org/apache/spark/util/JsonProtocolSuite.scala [hunk 1]
```diff
@@ -1103,6 +1107,9 @@ private[spark] object JsonProtocolSuite extends Assertions {
       sw.incBytesWritten(a + b + c)
       sw.incWriteTime(b + c + d)
       sw.incRecordsWritten(if (hasRecords) (a + b + c) / 100 else -1)
+      sw.incBlocksNotPushed(if (b > a) b - a else a - b)
+      sw.incBlocksCollided(if (b > a) b - a else a - b)
+      sw.incBlocksTooLate(if (b > a) b - a else a - b)
```
Comment (mridulm) [inline core/src/test/scala/org/apache/spark/util/JsonProtocolSuite.scala:1112]: nit: use `abs` instead ?

**Review by mridulm — COMMENTED** [review]
Took an initial pass through it. Thanks for working on it @thejdeep !

**Comment by mridulm** [comment]
+CC @Ngone51, @attilapiros

**Comment by mridulm** [comment]
Also +CC @dongjoon-hyun

**Comment by mridulm** [comment]
Can you fix the conflict @thejdeep ?

**Comment by pan3793** [comment]
Any update on this PR?

**Pull request closed without merging**